File: app/src/main/java/com/smile/karaokeplayer/Models/SongListSQLite.java
package com.smile.karaokeplayer.Models;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.net.Uri;
import android.util.Log;

import java.util.ArrayList;

/**
 * Created by lee on 09/13/2019.
 */
public class SongListSQLite extends SQLiteOpenHelper {

    private static final String TAG = "SongListSQLite";

    private static final String _id = new String("id");
    private static final String songName = new String("songName");
    private static final String filePath = new String("filePath");
    private static final String musicTrackNo = new String("musicTrackNo");
    private static final String musicChannel = new String("musicChannel");
    private static final String vocalTrackNo = new String("vocalTrackNo");
    private static final String vocalChannel = new String("vocalChannel");
    private static final String included = new String("included");

    private static final String dbName = new String("songDatabase.db");
    private static final String tableName = new String("songList");
    private static final String createTable = "create table if not exists " + tableName + " ("
            + _id + " INTEGER PRIMARY KEY AUTOINCREMENT, "
            + songName + " TEXT NOT NULL , "
            + filePath + " TEXT NOT NULL , "
            + musicTrackNo + " INTEGER , "
            + musicChannel + " INTEGER , "
            + vocalTrackNo + " INTEGER , "
            + vocalChannel + " INTEGER , "
            + included + " TEXT NOT NULL "
            + ");";

    private static final int dbVersion = 4;

    private static final int createAction = 0;
    private static final int readAction = 1;
    private static final int updateAction = 2;
    private static final int deleteAction = 3;

    private Context myContext;
    private SQLiteDatabase songDatabase;

    public SongListSQLite(Context context) {
        super(context, dbName,null,dbVersion);
        myContext = context;
        // the following statements is to create or update the database
        songDatabase = getWritableDatabase();
        closeDatabase();
    }

    @Override
    public void onCreate(SQLiteDatabase database) {
        database.execSQL(createTable);
    }

    @Override
    public void onUpgrade(SQLiteDatabase database , int oldVersion , int newVersion) {
        // database.execSQL("DROP TABLE IF EXISTS " + tableName);
        // onCreate(database);

        Log.d("TAG", "onUpgrade() is called.");

        try {
            String oldTableName = "playList";
            String sqlString = "ALTER TABLE " + oldTableName + " RENAME TO " + tableName;
            database.execSQL(sqlString);

            if (isColumnExist(database, included)) {
                Log.d(TAG, "included exists");
            } else {
                Log.d(TAG, "included does not exist");

                sqlString = "ALTER TABLE " + tableName + " ADD COLUMN " + included + " TEXT DEFAULT '1' NOT NULL ";
                database.execSQL(sqlString);
            }
        } catch (Exception ex) {
            Log.d("TAG", "Exception in onUpgrade().");
            ex.printStackTrace();
        }
    }

    private void openScoreDatabase() {

        if (songDatabase != null) {
            // already opened
            return;
        }
        try {
            songDatabase = getWritableDatabase();
        } catch (SQLException ex) {
            Log.d("TAG", "Open database exception.");
            ex.printStackTrace();
        }
    }

    private ContentValues getContentValues(SongInfo songInfo, int crudAction) {
        if (songInfo == null) {
            return null;
        }

        ContentValues contentValues = new ContentValues();
        if (crudAction != createAction) {
            contentValues.put(_id, songInfo.getId());
        }
        contentValues.put(songName, songInfo.getSongName());
        contentValues.put(filePath, songInfo.getFilePath());
        contentValues.put(musicTrackNo, songInfo.getMusicTrackNo());
        contentValues.put(musicChannel, songInfo.getMusicChannel());
        contentValues.put(vocalTrackNo, songInfo.getVocalTrackNo());
        contentValues.put(vocalChannel, songInfo.getVocalChannel());
        contentValues.put(included, songInfo.getIncluded());

        return contentValues;
    }

    private boolean isColumnExist(SQLiteDatabase database, String columnName) {
        boolean isExist = false;
        if (database != null) {
            try {
                // get only one record
                Cursor cur = database.query(tableName, null, null, null, null, null, null, "1");
                if (cur != null) {
                    int index = cur.getColumnIndex(columnName);
                    if (index != -1) {
                        // exist
                        isExist = true;
                    }
                }
                cur.close();
            } catch (SQLException ex) {
                ex.printStackTrace();
            }
        }

        return isExist;
    }

    private ArrayList<SongInfo> getSongInfoFromCursor(Cursor cur) {
        ArrayList<SongInfo> songList = new ArrayList<>();
        if (cur != null) {
            try {
                if (cur.moveToFirst()) {
                    do {
                        Integer id = cur.getInt(0);
                        String songName = cur.getString(1);
                        String filePath = cur.getString(2);
                        int musicTrackNo = cur.getInt(3);
                        int musicChannel = cur.getInt(4);
                        int vocalTrackNo = cur.getInt(5);
                        int vocalChannel = cur.getInt(6);
                        String included = cur.getString(7);
                        SongInfo songInfo = new SongInfo(id, songName, filePath, musicTrackNo, musicChannel, vocalTrackNo, vocalChannel, included);
                        songList.add(songInfo);
                    } while (cur.moveToNext());
                }
                cur.close();
            } catch (SQLException ex) {
                Log.d("TAG", "getSongInfoFromCursor() exception.");
                ex.printStackTrace();
            }
        }

        return songList;
    }

    public ArrayList<SongInfo> readSongList() {

        Log.d("TAG", "readSongList() is called.");

        ArrayList<SongInfo> songList = new ArrayList<>();

        openScoreDatabase();
        if (songDatabase != null) {
            try {
                // String sql = "select * from " + tableName + " order by " + _id + " asc";
                // Cursor cur = songDatabase.rawQuery(sql, new String[]{});
                Cursor cur = songDatabase.query(tableName, null, null, null, null, null, _id+" asc");
                songList = getSongInfoFromCursor(cur);
            } catch (SQLException ex) {
                Log.d("TAG", "readSongList() exception.");
                ex.printStackTrace();
            }
            closeDatabase();
        }

        return songList;
    }

    public ArrayList<SongInfo> readPlaylist() {

        Log.d("TAG", "readPlaylist() is called.");

        ArrayList<SongInfo> playlist = new ArrayList<>();

        openScoreDatabase();
        if (songDatabase != null) {
            try {
                // String sql = "select * from " + tableName + " order by " + _id + " asc";
                // Cursor cur = songDatabase.rawQuery(sql, new String[]{});
                Cursor cur = songDatabase.query(tableName, null, included + " = ?", new String[] {"1"}, null, null, _id+" asc");
                playlist = getSongInfoFromCursor(cur);
            } catch (SQLException ex) {
                Log.d("TAG", "readPlaylist() exception.");
                ex.printStackTrace();
            }
            closeDatabase();
        }

        return playlist;
    }

    public long addSongToSongList(final SongInfo songInfo) {

        long result = -1;

        if (songInfo == null) {
            return result;
        }

        ContentValues contentValues = getContentValues(songInfo, createAction);
        openScoreDatabase();
        if (songDatabase != null) {
            try {
                result = songDatabase.insert(tableName, null, contentValues);
            } catch (SQLException ex) {
                Log.d("TAG", "addSongToSongList() exception.");
                ex.printStackTrace();
            }
            closeDatabase();
        }

        return result;
    }

    public long updateOneSongFromSongList(final SongInfo songInfo) {

        long result = -1;

        if (songInfo == null) {
            return result;
        }

        ContentValues contentValues = getContentValues(songInfo, updateAction);
        String whereClause = _id + " = " + songInfo.getId();
        openScoreDatabase();
        if (songDatabase != null) {
            try {
                result = songDatabase.update(tableName, contentValues, whereClause, null);
            } catch (SQLException ex) {
                Log.d("TAG", "updateOneSongFromSongList() exception.");
                ex.printStackTrace();
            }
            closeDatabase();
        }

        return result;
    }

    public long deleteOneSongFromSongList(final SongInfo songInfo) {

        long result = -1;

        if (songInfo == null) {
            return result;
        }

        openScoreDatabase();
        if (songDatabase != null) {
            try {
                int id = songInfo.getId();
                String whereClause = _id + " = " + id;
                result = songDatabase.delete(tableName, whereClause,null);
            } catch (SQLException ex) {
                Log.d("TAG", "deleteOneSongFromSongList() exception.");
                ex.printStackTrace();
            }
            closeDatabase();
        }

        return result;
    }

    public void deleteAllSongList() {
            openScoreDatabase();
            if (songDatabase != null) {
                try {
                    songDatabase.delete(tableName, null,null);
                } catch (SQLException ex) {
                    Log.d("TAG", "deleteAllSongList() exception.");
                    ex.printStackTrace();
                }
                closeDatabase();
            }
    }

    public SongInfo findOneSongByContentUri(Uri contentUri) {
        SongInfo songInfo = null;
        String songFilePath = "";
        if (contentUri == null) {
            return songInfo;
        }
        songFilePath = contentUri.toString();

        openScoreDatabase();
        if (songDatabase != null) {
            try {
                String whereClause = filePath + " = " + '"' + songFilePath + '"';
                Cursor cur = songDatabase.query(tableName, null, whereClause, null, null, null, null);
                if (cur != null) {
                    if (cur.moveToFirst()) {
                        Integer id = cur.getInt(0);
                        String songName = cur.getString(1);
                        String filePath = cur.getString(2);
                        int musicTrackNo = cur.getInt(3);
                        int musicChannel = cur.getInt(4);
                        int vocalTrackNo = cur.getInt(5);
                        int vocalChannel = cur.getInt(6);
                        String included = cur.getString(7);
                        songInfo = new SongInfo(id, songName, filePath, musicTrackNo, musicChannel, vocalTrackNo, vocalChannel, included);
                    }
                    cur.close();
                }
            } catch (SQLException ex) {
                Log.d("TAG", "findOneSongByContentUri() exception.");
                ex.printStackTrace();
            }
            closeDatabase();
        }

        return songInfo;
    }

    public void closeDatabase() {
        if (songDatabase != null) {
            try {
                songDatabase.close();
                songDatabase = null;
            } catch (SQLException ex) {
                Log.d("TAG", "closeDatabase() exception.");
                ex.printStackTrace();
            }
        }
    }
}

File: app/src/main/java/com/smile/karaokeplayer/models/SongListSQLite.kt
package com.smile.karaokeplayer.models

import android.database.sqlite.SQLiteOpenHelper
import android.database.sqlite.SQLiteDatabase
import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.DatabaseUtils
import android.database.SQLException
import android.util.Log
import java.lang.Exception
import java.util.ArrayList

class SongListSQLite(myContext: Context) : SQLiteOpenHelper(
    myContext, dbName, null, dbVersion) {
    private var songDatabase: SQLiteDatabase?
    override fun onCreate(database: SQLiteDatabase) {
        database.execSQL(createTable)
    }

    override fun onUpgrade(database: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        // database.execSQL("DROP TABLE IF EXISTS " + tableName);
        // onCreate(database);
        Log.d(TAG, "onUpgrade() is called.")
        try {
            val oldTableName = "playList"
            var sqlString = "ALTER TABLE $oldTableName RENAME TO $tableName"
            database.execSQL(sqlString)
            if (isColumnExist(database, included)) {
                Log.d(TAG, "included exists")
            } else {
                Log.d(TAG, "included does not exist")
                sqlString =
                    "ALTER TABLE $tableName ADD COLUMN $included TEXT DEFAULT '1' NOT NULL "
                database.execSQL(sqlString)
            }
        } catch (ex: Exception) {
            Log.e(TAG, "Exception in onUpgrade().")
            ex.printStackTrace()
        }
    }

    private fun openScoreDatabase() {
        songDatabase?.let {
            // already opened
            return
        }
        try {
            songDatabase = writableDatabase
        } catch (ex: SQLException) {
            Log.e(TAG, "Open database exception.")
            ex.printStackTrace()
        }
    }

    private fun getContentValues(songInfo: SongInfo?, crudAction: Int): ContentValues? {
        if (songInfo == null) {
            return null
        }
        val contentValues = ContentValues()
        if (crudAction != createAction) {
            contentValues.put(mId, songInfo.id)
        }
        contentValues.put(songName, songInfo.songName)
        contentValues.put(filePath, songInfo.filePath)
        contentValues.put(musicTrackNo, songInfo.musicTrackNo)
        contentValues.put(musicChannel, songInfo.musicChannel)
        contentValues.put(vocalTrackNo, songInfo.vocalTrackNo)
        contentValues.put(vocalChannel, songInfo.vocalChannel)
        contentValues.put(included, songInfo.included)
        return contentValues
    }

    private fun isColumnExist(database: SQLiteDatabase?, columnName: String): Boolean {
        if (database == null) {
            return false
        }
        var isExist = false
        try {
            // get only one record
            val cur = database.query(tableName, null, null, null, null, null, null, "1")
            cur?.run {
                val index = getColumnIndex(columnName)
                if (index != -1) {
                    // exist
                    isExist = true
                }
                close()
            }
        } catch (ex: SQLException) {
            ex.printStackTrace()
        }
        return isExist
    }

    private fun getSongInfoFromCursor(cur: Cursor?): ArrayList<SongInfo> {
        val songList = ArrayList<SongInfo>()
        cur?.run {
            try {
                if (moveToFirst()) {
                    do {
                        val id = getInt(0)
                        val songName = getString(1)
                        val filePath = getString(2)
                        val musicTrackNo = getInt(3)
                        val musicChannel = getInt(4)
                        val vocalTrackNo = getInt(5)
                        val vocalChannel = getInt(6)
                        val included = getString(7)
                        val songInfo = SongInfo(id, songName, filePath, musicTrackNo, musicChannel,
                                vocalTrackNo, vocalChannel, included)
                        songList.add(songInfo)
                    } while (moveToNext())
                }
                close()
            } catch (ex: SQLException) {
                Log.e(TAG, "getSongInfoFromCursor() exception.")
                ex.printStackTrace()
            }
        }
        return songList
    }

    fun recordsOfPlayList() : Long {
        openScoreDatabase()
        songDatabase?.let {
            return DatabaseUtils.queryNumEntries(it, tableName)
        }
        return 0
    }
    fun readPlayList(): ArrayList<SongInfo> {
        Log.d(TAG, "readSongList() is called.")
        return readPlaylist(false)
    }

    fun readPlaylist(isIncluded: Boolean): ArrayList<SongInfo> {
        Log.d(TAG, "readPlaylist().isIncluded = $isIncluded")
        var playlist = ArrayList<SongInfo>()
        openScoreDatabase()
        songDatabase?.let {
            try {
                // String sql = "select * from " + tableName + " order by " + _id + " asc";
                // Cursor cur = songDatabase.rawQuery(sql, new String[]{});
                var selectionArgs: Array<String>? = null
                var selection: String? = null
                if (isIncluded) {
                    selectionArgs = arrayOf("1")
                    selection = "$included = ?"
                }
                val cur = it.query(tableName, null, selection, selectionArgs,
                        null, null,"$mId asc")
                playlist = getSongInfoFromCursor(cur)
            } catch (ex: SQLException) {
                Log.e(TAG, "readPlaylist() exception.")
                ex.printStackTrace()
            }
        }
        return playlist
    }

    fun addSongToSongList(songInfo: SongInfo?): Long {
        var result: Long = -1
        if (songInfo == null) {
            return result
        }
        openScoreDatabase()
        if (findOneSongByUriString(songInfo.filePath) != null ) return result   // already in database
        val contentValues = getContentValues(songInfo, createAction)
        songDatabase?.let {
            try {
                result = it.insert(tableName, null, contentValues)
            } catch (ex: SQLException) {
                Log.e(TAG, "addSongToSongList() exception.")
                ex.printStackTrace()
            }
        }
        return result
    }

    fun updateOneSongFromSongList(songInfo: SongInfo?): Long {
        var result: Long = -1
        if (songInfo == null) {
            return result
        }
        val contentValues = getContentValues(songInfo, updateAction)
        val whereClause = mId + " = " + songInfo.id
        openScoreDatabase()
        songDatabase?.let {
            try {
                result = it.update(tableName, contentValues, whereClause, null).toLong()
            } catch (ex: SQLException) {
                Log.e(TAG, "updateOneSongFromSongList() exception.")
                ex.printStackTrace()
            }
        }
        return result
    }

    fun deleteOneSongFromSongList(songInfo: SongInfo?): Long {
        var result: Long = -1
        if (songInfo == null) {
            return result
        }
        openScoreDatabase()
        songDatabase?.let {
            try {
                val id = songInfo.id
                val whereClause = "$mId = $id"
                result = it.delete(tableName, whereClause, null).toLong()
            } catch (ex: SQLException) {
                Log.e(TAG, "deleteOneSongFromSongList() exception.")
                ex.printStackTrace()
            }
        }
        return result
    }

    fun deleteAllSongList() {
        openScoreDatabase()
        songDatabase?.let {
            try {
                it.delete(tableName, null, null)
            } catch (ex: SQLException) {
                Log.e(TAG, "deleteAllSongList() exception.")
                ex.printStackTrace()
            }
        }
    }

    fun findOneSongByUriString(uriString: String?): SongInfo? {
        if (uriString == null || uriString.isEmpty()) {
            return null
        }
        openScoreDatabase()
        var songInfo: SongInfo? = null
        songDatabase?.let {
            try {
                val whereClause = "$filePath = \"$uriString\""
                val cur = it.query(tableName, null, whereClause, null, null, null, null)
                cur?.run {
                    if (moveToFirst()) {
                        val id = getInt(0)
                        val songName = getString(1)
                        val filePath = getString(2)
                        val musicTrackNo = getInt(3)
                        val musicChannel = getInt(4)
                        val vocalTrackNo = getInt(5)
                        val vocalChannel = getInt(6)
                        val included = getString(7)
                        songInfo = SongInfo(id, songName, filePath, musicTrackNo,
                                musicChannel, vocalTrackNo, vocalChannel, included)
                    }
                    close()
                }
            } catch (ex: SQLException) {
                Log.e(TAG, "findOneSongByContentUri() exception.")
                ex.printStackTrace()
            }
        }
        return songInfo
    }

    fun closeDatabase() {
        songDatabase?.let {
            try {
                it.close()
            } catch (ex: SQLException) {
                Log.e(TAG, "closeDatabase() exception.")
                ex.printStackTrace()
            }
        }
        songDatabase = null
    }

    companion object {
        private const val TAG = "SongListSQLite"
        private const val mId: String = "id"
        private const val songName: String = "songName"
        private const val filePath: String = "filePath"
        private const val musicTrackNo: String = "musicTrackNo"
        private const val musicChannel: String = "musicChannel"
        private const val vocalTrackNo: String = "vocalTrackNo"
        private const val vocalChannel: String = "vocalChannel"
        private const val included: String = "included"
        private const val dbName: String = "songDatabase.db"
        private const val tableName: String = "songList"
        private const val createTable = ("create table if not exists " + tableName + " ("
                + mId + " INTEGER PRIMARY KEY AUTOINCREMENT, "
                + songName + " TEXT NOT NULL , "
                + filePath + " TEXT NOT NULL , "
                + musicTrackNo + " INTEGER , "
                + musicChannel + " INTEGER , "
                + vocalTrackNo + " INTEGER , "
                + vocalChannel + " INTEGER , "
                + included + " TEXT NOT NULL "
                + ");")
        private const val dbVersion = 4
        private const val createAction = 0
        private const val readAction = 1
        private const val updateAction = 2
        private const val deleteAction = 3
    }

    init {
        // the following statements is to create or update the database
        songDatabase = writableDatabase
        // closeDatabase();
    }
}